tests/testsupport/broker.py

```diff
@@ -0,0 +1,62 @@
+import socket
+
+import pytest
+
+
+class FakeBroker:
+    def __init__(self):
+        # Bind to "localhost" for maximum performance, as described in:
+        # http://docs.python.org/howto/sockets.html#ipc
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        sock.settimeout(30)
+        sock.bind(("localhost", 1888))
+        sock.listen(1)
+
+        self._sock = sock
+        self._conn = None
+
+    def start(self):
+        if self._sock is None:
+            raise ValueError('Socket is not open')
+
+        (conn, address) = self._sock.accept()
+        conn.settimeout(10)
+        self._conn = conn
+
+    def finish(self):
+        if self._conn is not None:
+            self._conn.close()
+            self._conn = None
+
+        if self._sock is not None:
+            self._sock.close()
+            self._sock = None
+
+    def expect_packet(self, packet_expected):
+        if self._conn is None:
+            raise ValueError('Connection is not open')
+
+        packet_in = self._conn.recv(len(packet_expected))
+        if not packet_in:
```

Comment:
That's not really an assertion (i.e. thing that we are testing) rather an unexpected event.

This is just a first version. There needs to be **a lot** of refactoring to make paho more testable. I want to be using a fake socket for most of the tests in the future, but paho currently lacks the dependency injection functionality that would be required.